source/opt/fix_storage_class.cpp

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "fix_storage_class.h"
+
+#include "source/opt/instruction.h"
+#include "source/opt/ir_context.h"
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status FixStorageClass::Process() {
+  bool modified = false;
+
+  get_module()->ForEachInst([this, &modified](Instruction* inst) {
+    if (FixInstruction(inst)) {
+      modified = true;
+    }
+  });
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool FixStorageClass::FixInstruction(Instruction* inst) {
+  switch (inst->opcode()) {
+    case SpvOpAccessChain:
+    case SpvOpPtrAccessChain:
+    case SpvOpInBoundsAccessChain:
+    case SpvOpCopyObject:
+    case SpvOpPhi:
+      return FixAddressInstruction(inst, 0);
+    case SpvOpSelect:
+      return FixAddressInstruction(inst, 1);
+    case SpvOpFunctionCall:
+      // If the same function is called twice with different storage classes for
+      // the parameters, then it is impossible to fix it.  If function calls are
+      // a problem, then should be inlined, and then run this pass.
+      return false;
+    case SpvOpImageTexelPointer:
+    case SpvOpLoad:
+    case SpvOpStore:
+    case SpvOpCopyMemory:
+    case SpvOpCopyMemorySized:
+      // Nothing to change for these opcode.  The result type is the same
+      // regardless of the storage class of the operand.
+      return false;
+    default:
+      // If we get here, then it is expected that the instruction cannot
+      // possibly have a pointer result or operand.
+      return false;
+  }
+}
+
+bool FixStorageClass::FixAddressInstruction(Instruction* inst,
+                                            uint32_t in_operand_idx) {
+  assert((inst->opcode() == SpvOpAccessChain ||
+          inst->opcode() == SpvOpPtrAccessChain ||
+          inst->opcode() == SpvOpInBoundsAccessChain ||
+          inst->opcode() == SpvOpCopyObject) &&
```

Comment:
`|| inst->opcode() == SpvOpPhi` ?

Reply:
Yes, I will need to fix this.  I still have to add more tests as well.